### PR TITLE
Adds an abstraction layer to hide the underlying usage of Microsoft Data Encryption(MDE) library

### PR DESCRIPTION
**Pull request by simplynaveen20**
This PR adds in an abstraction layer to hide the underlying dependency on Microsoft Data Encryption library. The abstraction is limited to only the classes which are currently exposed to public methods of client side encryption SDK.

Also added module-info for encryption  project

.NET PR - https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2906

- We have moved away from AAP EncryptionKeyStoreProvider  during client creation , and created our own EncryptionKeyWrapProvider.
- This PR also introduced the wrapper implementation of AzureKeyVaultKeyStoreProvider in a cosmos lib wrapper as AzureKeyVaultKeyWrapProvider.

This PR is prerequisite for AAP shading jar PR https://github.com/Azure/azure-sdk-for-java/pull/26738

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * @param encryptionKeyId                 - Complete path of an asymmetric key in Azure Key Vault
+     * @param cosmosKeyEncryptionKeyAlgorithm - Asymmetric Key Encryption Algorithm
+     * @param encryptedKey                    - Encrypted Data Encryption Key
+     * @return Plain text data encryption key
+     */
+    @Override
+    public byte[] unwrapKeyAsync(String encryptionKeyId, String cosmosKeyEncryptionKeyAlgorithm, byte[] encryptedKey) {
+        try {
+            if (!KeyEncryptionKeyAlgorithm.RSA_OAEP.equals(cosmosKeyEncryptionKeyAlgorithm)) {
+                throw new IllegalArgumentException("The specified KeyEncryptionAlgorithm is not supported. Please " +
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java:67]: should we put the algorithm name in the message?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.encryption.mdesupport;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MdeSupportBridgeHelpers {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java:9]: Shouldn't we add `@Warning` annotation to this class and all its public methods ?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public static final class EncryptionKeyWrapProviderHelper {
+        static {
+            ensureClassLoaded(EncryptionKeyWrapProvider.class);
+        }
+        private static EncryptionKeyWrapProviderAccessor accessor;
+
+        private EncryptionKeyWrapProviderHelper() {}
+
+        public static void setPEncryptionKeyWrapProviderAccessor(final EncryptionKeyWrapProviderAccessor newAccessor) {
+            if (accessor != null) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java:20]: should check == null here?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyWrapProvider.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.encryption.mdesupport;
+
+import com.microsoft.data.encryption.cryptography.MicrosoftDataEncryptionException;
+
+public abstract class EncryptionKeyWrapProvider {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyWrapProvider.java:8]: Why do we need abstract class here? Can't we make this as an interface?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * Encrypts a DEK with the RSA encryption algorithm using the asymmetric key specified by the key path.
+     *
+     * @param encryptionKeyId                 - Complete path of an asymmetric key in Azure Key Vault
+     * @param cosmosKeyEncryptionKeyAlgorithm - Asymmetric Key Encryption Algorithm
+     * @param key                             - Plain text data encryption key
+     * @return Encrypted data encryption key
+     */
+    @Override
+    public byte[] wrapKeyAsync(String encryptionKeyId, String cosmosKeyEncryptionKeyAlgorithm, byte[] key) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java:86]: another question is why the method is async suffix? seems like there is no async operation in the method.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyStoreProviderImpl.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+     *        The key Id tells the provider where to find the key.
+     * @param allowEnclaveComputations
+     *        Indicates whether the key encryption key supports enclave computations.
+     * @return The signature of the key encryption key metadata.
+     * @throws MicrosoftDataEncryptionException
+     *         on error
+     */
+    @Override
+    public byte[] sign(String encryptionKeyId, boolean allowEnclaveComputations) {
+        return new byte[0]; // "The Sign operation is not supported. " ;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyStoreProviderImpl.java:78]: if not supported, should we throw exception instead?

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyStoreProviderImpl.java:78]: +1 here

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java (hunk 1 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java:67]: This is if something change in the AAP which support more algorith , cosmos will have only RSA_OAEP as of now , same as .NET https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5c01524de44ea0104223f6d494edeb98ac30a44c/Microsoft.Azure.Cosmos.Encryption/src/MdeSupport/AzureKeyVaultKeyWrapProvider.cs#L52

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java (hunk 2 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/MdeSupportBridgeHelpers.java:9]: Done, btw in cosmos core project we dont have warning for BridgeHelpers, something we should do there as well

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyWrapProvider.java (hunk 4 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyWrapProvider.java:17]: Once AAP java support catching , we can provide default implementation here and give user the way to override same as .NET https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5c01524de44ea0104223f6d494edeb98ac30a44c/Microsoft.Azure.Cosmos.Encryption/src/MdeSupport/EncryptionKeyWrapProvider.cs#L32

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java (hunk 5 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/AzureKeyVaultKeyWrapProvider.java]: good catch , removed async

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyStoreProviderImpl.java (hunk 6 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/mdesupport/EncryptionKeyStoreProviderImpl.java:78]: That's what i started with but AAP KeyEncryptionKey.java constructor calling this method , if this would have been async/lazy like .net it would have make sense to throw whenever called as this is not getting called after KeyEncryptionKey initialization.
But in Java we have two choice either return null and empty array (no signing) , so i choose empty array , verified with .NET encryption  dev

**Review by kushagraThapar — APPROVED** [review]

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos-encryption/pom.xml [hunk 7]
```diff
@@ -36,11 +36,20 @@ Licensed under the MIT License.
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <jacoco.min.linecoverage>0.10</jacoco.min.linecoverage>
-    <jacoco.min.branchcoverage>0.10</jacoco.min.branchcoverage>
+    <jacoco.min.branchcoverage>0.05</jacoco.min.branchcoverage>
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/pom.xml:39]: why the change here?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncContainer.java [hunk 8]
```diff
@@ -1,19 +1,19 @@
...
 import com.azure.cosmos.CosmosBridgeInternal;
 import com.azure.cosmos.CosmosException;
-import com.azure.cosmos.encryption.implementation.Constants;
-import com.azure.cosmos.encryption.implementation.CosmosResponseFactory;
-import com.azure.cosmos.encryption.implementation.EncryptionProcessor;
-import com.azure.cosmos.encryption.implementation.EncryptionUtils;
-import com.azure.cosmos.encryption.models.EncryptionModelBridgeInternal;
-import com.azure.cosmos.encryption.models.SqlQuerySpecWithEncryption;
-import com.azure.cosmos.encryption.util.Beta;
+import com.azure.encryption.cosmos.util.Beta;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncContainer.java:10]: the order seems a little off

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncDatabase.java [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
 
-package com.azure.cosmos.encryption;
+package com.azure.encryption.cosmos;
 
 import com.azure.cosmos.CosmosAsyncClientEncryptionKey;
 import com.azure.cosmos.CosmosAsyncContainer;
 import com.azure.cosmos.CosmosAsyncDatabase;
-import com.azure.cosmos.encryption.keyprovider.EncryptionKeyWrapProvider;
-import com.azure.cosmos.encryption.keyprovider.KeyProviderBridgeHelpers;
+import com.azure.encryption.cosmos.keyprovider.EncryptionKeyWrapProvider;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncDatabase.java:9]: same here: the order seems

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncContainer.java (hunk 8 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/encryption/cosmos/CosmosEncryptionAsyncContainer.java]: done

File: sdk/cosmos/azure-cosmos-encryption/pom.xml (hunk 7 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/pom.xml:39]: branch coverage is .05 currently, as we rely heavily on E2E test
